### PR TITLE
Add environ to bootstrap worker

### PR DESCRIPTION
**Pull request by nvinuesa**
This patch is a preamble to the work being done to remove usages of the space/subnet domain from the legacy state. We need the environ because in order to init the API host port at bootstrap we'll need to retrieve the instance address from the environ in the bootstrap worker.

It must be noted that this work tends to move some logic from agent bootstrap to the bootstrap worker.

## Checklist

- [X] Code style: imports ordered, good names, simple structure, etc
- [X] Comments saying why design decisions were made
- [X] Go unit tests, with comments saying what you're testing
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

Nothing is wired yet:

```
go test github.com/juju/juju/internal/worker/bootstrap/... -gocheck.v        
```

also:
```
juju bootstrap lxd c
juju add-model m
juju deploy ubuntu
```

## Links

**Jira card:** JUJU-5107

### CODE REVIEW

**Review by SimonRichardson — CHANGES_REQUESTED** [review]
Run `go test -v ./cmd/jujud/agent/machine -check.v -check.f=ManifoldsSuite` and you'll see failures.

**Comment by nvinuesa** [comment]
@SimonRichardson I added a new method to get an environ from the worker, instead of having it as an input; as discussed. Also bootstrapping and deploying is part of QA

**Review by SimonRichardson — APPROVED** [review]
Lovely, just close out the test failures and this is happy to go.

File: internal/worker/bootstrap/worker.go [hunk 1]
```diff
@@ -250,6 +290,51 @@ func (w *bootstrapWorker) bootstrapParams(ctx context.Context, dataDir string) (
...
+			return nil, errors.Trace(err)
+		}
+		cloudCredential := cloud.NewNamedCredential(
+			credentialValue.Label,
+			credentialValue.AuthType(),
+			credentialValue.Attributes(),
+			credentialValue.Revoked,
+		)
+		cred = &cloudCredential
+	}
```
Comment (SimonRichardson) [inline internal/worker/bootstrap/worker.go:314]: Also can we put that into another method, just to make readability nicer. `getEnvironCredential`.

Comment (nvinuesa) [reply internal/worker/bootstrap/worker.go]: done

File: internal/worker/bootstrap/worker.go [hunk 2]
```diff
@@ -156,6 +176,26 @@ func (w *bootstrapWorker) loop() error {
...
+	addresses, err = w.cfg.BootstrapAddressesFunc(ctx, env)
+	if err != nil && !errors.Is(err, errors.NotSupported) {
+		return errors.Trace(err)
+	}
+	if errors.Is(err, errors.NotSupported) {
+		addresses = network.NewMachineAddresses([]string{"localhost"}).AsProviderAddresses()
+	}
+	// NOTE(nvinuesa): This is only temporary, its usage will be
+	// implemented in a future patch.
+	_ = addresses
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply internal/worker/bootstrap/worker.go]: done

File: internal/worker/bootstrap/manifold.go [hunk 3]
```diff
@@ -80,6 +98,31 @@ type HTTPClient interface {
...
+// InstanceListener (this is the case of CAAS for the moment), then we
+// return the hard-coded 'localhost' address.
+func BootstrapAddresses(ctx context.Context, env environs.Environ) (network.ProviderAddresses, error) {
+	callCtx := envcontext.WithoutCredentialInvalidator(ctx)
+	instanceLister, ok := env.(environs.InstanceLister)
+	if !ok {
+		// NOTE(nvinuesa): This will happen on caas providers where
+		// the environ does not implement InstanceListener, in this
+		// we return "localhost" as address.
+		return network.NewMachineAddresses([]string{"localhost"}).AsProviderAddresses(), nil
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply internal/worker/bootstrap/manifold.go]: done

File: internal/worker/bootstrap/manifold.go [hunk 4]
```diff
@@ -80,6 +98,31 @@ type HTTPClient interface {
...
+	}
+	// TODO(nvinuesa): which instanceID to use?
+	instances, err := instanceLister.Instances(callCtx, []instance.Id{"0"})
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply internal/worker/bootstrap/manifold.go]: great! I did that

File: internal/worker/bootstrap/worker.go [hunk 5]
```diff
@@ -202,20 +253,20 @@ func (w *bootstrapWorker) seedAgentBinary(ctx context.Context, dataDir string) (
 	return cleanup, nil
 }
 
-func (w *bootstrapWorker) seedControllerCharm(ctx context.Context, dataDir string) error {
+func (w *bootstrapWorker) seedControllerCharm(ctx context.Context, dataDir string) (instancecfg.StateInitializationParams, error) {
```
Comment (SimonRichardson) [inline internal/worker/bootstrap/worker.go:256]: Just move `bootstrapParams` out of `seedControllerCharm`.

Comment (nvinuesa) [reply internal/worker/bootstrap/worker.go]: done

File: internal/worker/bootstrap/worker.go [hunk 6]
```diff
@@ -152,7 +174,18 @@ func (w *bootstrapWorker) loop() error {
...
+		return errors.Annotatef(err, "getting bootstrap params")
+	}
+	bootstrapArgs, err := w.seedControllerCharm(ctx, dataDir, args)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Retrieve bootstrap addresses for setting API host ports.
+	_, err = w.getBoostrapAddresses(ctx, bootstrapArgs.BootstrapMachineInstanceId)
+	if err != nil {
```
Comment (SimonRichardson) [inline internal/worker/bootstrap/worker.go:188]: Excellent.

**Review by SimonRichardson — APPROVED** [review]

**Pull request merged**